// udoo/lib/constant.js
/*
 * lib/constant.js
 */

'use strict';

var _ = require('lodash');

exports._PATH_FORMAT = {
  DIRECTION: '/sys/class/gpio/gpio%d/direction',
  VALUE: '/sys/class/gpio/gpio%d/value'
};

exports._PIN_MAPPING = {
  2: 20,
  3: 16,
  4: 17,
  5: 18,
  6: 41,
  7: 42,
  8: 21,
  9: 19,
  10: 1,
  11: 9,
  12: 3,
  13: 40,
  14: 150,
  15: 162,
  16: 160,
  17: 161,
  18: 158,
  19: 159,
  20: 92,
  21: 85,
  22: 123,
  23: 124,
  24: 125,
  25: 126,
  26: 127,
  27: 133,
  28: 134,
  29: 135,
  30: 136,
  31: 137,
  32: 138,
  33: 139,
  34: 140,
  35: 141,
  36: 142,
  37: 143,
  38: 54,
  39: 205,
  40: 32,
  41: 35,
  42: 34,
  43: 33,
  44: 101,
  45: 144,
  46: 145,
  47: 89,
  48: 105,
  49: 104,
  50: 57,
  51: 56,
  52: 55,
  CANRX: 7,
  CANTX: 8,
  D22: 123,
  D23: 124,
  D24: 125,
  D25: 126,
  D26: 127,
  D27: 133,
  D28: 134,
  D29: 135,
  D30: 136,
  D31: 137,
  D32: 138,
  D33: 139,
  D34: 140,
  D35: 141,
  D36: 142,
  D37: 143,
  D38: 54,
  D39: 205,
  D40: 32,
  D41: 35,
  D42: 34,
  D43: 33,
  D44: 101,
  D45: 144,
  D46: 145,
  D47: 89,
  D48: 105,
  D49: 104,
  D50: 57,
  D51: 56,
  D52: 55,
  PWM2: 20,
  PWM3: 16,
  PWM4: 17,
  PWM5: 18,
  PWM6: 41,
  PWM7: 42,
  PWM8: 21,
  PWM9: 19,
  PWM10: 1,
  PWM11: 9,
  PWM12: 3,
  PWM13: 40,
  RX1: 159,
  RX2: 161,
  RX3: 162,
  SDA: 92,
  SCL: 85,
  TX1: 158,
  TX2: 160,
  TX3: 150
};

exports.PIN_MODE = {
  INPUT: 'in',
  OUTPUT: 'out'
};

exports.PIN_MODE_INVERT = _.invert(exports.PIN_MODE);
